narwhal/consensus/src/bullshark.rs

```diff
@@ -316,13 +275,114 @@ impl Bullshark {
+                leader = prev_leader;
+            } else {
+                self.metrics
+                    .leader_election
+                    .with_label_values(&["no_path", authority.hostname()])
+                    .inc();
+            }
+        }
+
+        // Now just report all the found leaders
+        let committee = self.committee.clone();
+        let metrics = self.metrics.clone();
+
+        to_commit.iter().for_each(|certificate| {
+            let authority = committee.authority(&certificate.origin()).unwrap();
+
+            metrics
+                .leader_election
+                .with_label_values(&["committed", authority.hostname()])
+                .inc();
+        });
+
+        to_commit
+    }
+
+    /// Checks if there is a path between two leaders.
+    fn linked(&self, leader: &Certificate, prev_leader: &Certificate, dag: &Dag) -> bool {
+        let mut parents = vec![leader];
+        for r in (prev_leader.round()..leader.round()).rev() {
+            parents = dag
+                .get(&r)
+                .expect("We should have the whole history by now")
+                .values()
+                .filter(|(digest, _)| {
+                    parents
+                        .iter()
+                        .any(|x| x.header().parents().contains(digest))
+                })
+                .map(|(_, certificate)| certificate)
+                .collect();
+        }
+        parents.contains(&prev_leader)
+    }
+
+    // When the provided `reputation_scores` are "final" for the current schedule window, then we
+    // create the new leader swap table and update the leader schedule to use it. Otherwise we do
+    // nothing. If the schedule has been updated then true is returned.
+    fn update_leader_schedule(
+        &mut self,
+        leader_round: Round,
+        reputation_scores: &ReputationScores,
+    ) -> bool {
+        // Do not perform any update if the feature is disabled
+        if self.protocol_config.narwhal_new_leader_election_schedule()
+            && reputation_scores.final_of_schedule
+        {
+            // create the new swap table and update the scheduler
+            self.leader_schedule
+                .update_leader_swap_table(LeaderSwapTable::new(
+                    &self.committee,
+                    leader_round,
+                    reputation_scores,
+                ));
+
+            return true;
+        }
+        false
+    }
+
     fn report_leader_on_time_metrics(&mut self, certificate_round: Round, state: &ConsensusState) {
         if certificate_round > self.max_inserted_certificate_round
             && certificate_round % 2 == 0
             && certificate_round > 2
         {
             let previous_leader_round = certificate_round - 2;
-            let authority = Self::leader_authority(&self.committee, previous_leader_round);
+
+            // This metric will not be accurate anymore on the border when we do change schedules,
```

Comment:
This sentence is not very clear to me. Does it mean the authority label can be inaccurate sometimes?

Reply:
Exactly, since we report the hit/miss for the previous round/leader, when we do update the schedule we do calculate the leader based on the new schedule which _might_ be different now.

Reply:
Can you clarify that it is the authority label (as opposed to the total count) that can be inaccurate?

Reply:
Absolutely!